README.md
# klutz

**K**lean and **ut**ili**z**able library of common data types and useful functions. Also,

**klutz** <span style="color:darkgray">| kləts |</span>

noun <span style="color:darkgray">_informal, mainly North American_</span>

a clumsy, awkward, or foolish person.

**<span style="font-variant:small-caps;">Origin</span>**

1930s: from Yiddish **klots** ‘wooden block’.

See the [Wiki](https://github.com/sobchak-security/klutz/wiki) for more information.


RELEASE_NOTES.md
# Release Notes

## v0.0.1

* Add unmarshaling support for `zap.Logger` configuration.

pkg/log/log_test.go
// Copyright (c) 2023 Remo Ronca 106963724+sobchak-security@users.noreply.github.com
// MIT License

package log

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"testing"
	"time"

	"go.uber.org/zap"
)

func testTmpFile(t *testing.T, content string) (*os.File, func()) {
	t.Helper()

	f, err := os.CreateTemp("", "test_*")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := f.Write([]byte(content)); err != nil {
		t.Error(err)
	}
	return f, func() {
		if err := f.Close(); err != nil {
			t.Error(err)
		}
		if err := os.Remove(f.Name()); err != nil {
			t.Error(err)
		}
	}
}

func TestAdHoc(t *testing.T) {
	conf := []byte(`{
		"disableStacktrace": true,
		"encoding": "console",
		"level": "info",
		"outputPaths": ["stderr"],
		"encoderConfig": {
		  "levelKey": "level",
		  "levelEncoder": "capital",
		  "messageKey": "short_message",
		  "timeEncoder": "short",
		  "timeKey": "timestamp"
		},
		"initialFields": {
		  "host": "${HOSTNAME}"
		}
	}`)
	var m map[string]interface{}
	var cfg zap.Config

	if err := json.Unmarshal(conf, &m); err != nil {
		t.Fatal(err)
	}
	if err := (*ConfigWrapper)(&cfg).Unmarshal(m); err != nil {
		t.Fatal(err)
	}

	logger, err := cfg.Build()
	if err != nil {
		t.Fatal(err)
	}

	logger.Debug("not included with info level")
	logger.Warn("a warning")

	// t.Error("intentional")
}

func TestUnmarshal(t *testing.T) {
	type args struct {
		m map[string]interface{}
	}
	tests := []struct {
		name    string
		cfg     zap.Config
		conf    func(logfile *os.File) string
		args    args
		logF    func(*zap.Logger)
		wantLog string
		wantErr bool
	}{
		{
			name: "success: custom configuration with short time encoder",
			cfg:  zap.Config{},
			conf: func(logfile *os.File) string {
				return fmt.Sprintf(`{
				  "disableCaller": false,
				  "disableStacktrace": false,
				  "encoding": "console",
				  "errorOutputPaths": [
					"stderr"
				  ],
				  "level": "warn",
				  "outputPaths": [
					"stdout",
					%q
				  ],
				  "encoderConfig": {
					"callerEncoder": "full",
					"durationEncoder": "string",
					"levelEncoder": "capital",
					"levelKey": "level",
					"messageKey": "short_message",
		
					"timeEncoder": %q,
					"timeKey": "timestamp"
				  },
				  "initialFields": {
					"host": "${HOSTNAME}",
					"version": "1.1"
				  }
			  }`, logfile.Name(), ConfigKeyTimeEncoderShort)
			},
			logF: func(l *zap.Logger) {
				l.Info("info")
				l.Error("error", zap.Duration("_duration", 3*time.Hour+5*time.Minute+7*time.Second))
			},
			wantLog: `\d\d-\d\d-\d\d \d\d:\d\d:\d\d\s+ERROR\s+error\s+{"host": ".+", "version": "1.1", "_duration": "3h5m7s"}`,
		},
		{
			name: "success: custom configuration with int64 time encoder",
			cfg:  zap.Config{},
			conf: func(logfile *os.File) string {
				return fmt.Sprintf(`{
				  "disableCaller": true,
				  "disableStacktrace": true,
				  "encoding": "console",
				  "level": "info",
				  "outputPaths": [
					%q
				  ],
				  "encoderConfig": {
					"durationEncoder": "string",
					"levelKey": "level",
					"messageKey": "short_message",
		
					"timeEncoder": %q,
					"timeKey": "timestamp"
				  }
			  }`, logfile.Name(), ConfigKeyTimeEncoderInt64Seconds)
			},
			logF: func(l *zap.Logger) {
				l.Info("info")
			},
			wantLog: `\d{8}	info`,
		},
		{
			name: "failure: invalid level",
			cfg:  zap.Config{},
			args: args{
				m: map[string]interface{}{
					"level": "invalid",
				},
			},
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			f, closer := testTmpFile(t, "")
			defer closer()

			m := tt.args.m
			if m == nil {
				if err := json.Unmarshal([]byte(tt.conf(f)), &m); err != nil {
					t.Fatal(err)
				}
			}
			if err := (*ConfigWrapper)(&tt.cfg).Unmarshal(m); err != nil {
				if !tt.wantErr {
					t.Errorf("UnmarshalConfig() error = %v, wantErr %v", err, true)
				}
				return
			}

			logger, err := tt.cfg.Build()
			if err != nil {
				t.Fatal(err)
			}
			tt.logF(logger)

			b, err := os.ReadFile(f.Name())
			if err != nil {
				t.Fatal(err)
			}
			ok, err := regexp.MatchString(tt.wantLog, string(b))
			if err != nil {
				t.Errorf("UnmarshalConfig(): MatchString failed %v", err)
			}
			if !ok {
				t.Errorf("UnmarshalConfig() sample log output: %q, want %q",
					string(b), tt.wantLog)
			}
			// t.Error("intentional")
		})
	}
}

func TestDevConfig(t *testing.T) {
	f, closer := testTmpFile(t, "")
	defer closer()

	cfg := DevConfig()
	cfg.OutputPaths = append(cfg.OutputPaths, f.Name())

	l, err := cfg.Build()
	if err != nil {
		t.Fatal(err)
	}

	// 2023-04-21T09:41:58.212+0200	DEBUG	log/log_test.go:179	debug
	wantLog := `\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d\.\d{3}[+|-]\d{4}\s+DEBUG\s+log/log_test\.go:\d+\s+debug\s*$`

	l.Debug("debug")

	b, err := os.ReadFile(f.Name())
	if err != nil {
		t.Fatal(err)
	}
	ok, err := regexp.MatchString(wantLog, string(b))
	if err != nil {
		t.Errorf("UnmarshalConfig(): MatchString failed %v", err)
	}
	if !ok {
		t.Errorf("UnmarshalConfig() sample log output: %q, want %q",
			string(b), wantLog)
	}

	// t.Error("intentional")
}

func TestStdConfig(t *testing.T) {
	f, closer := testTmpFile(t, "")
	defer closer()

	cfg := StdConfig()
	cfg.OutputPaths = append(cfg.OutputPaths, f.Name())

	l, err := cfg.Build()
	if err != nil {
		t.Fatal(err)
	}

	// 2023-04-21T09:41:58.212+0200	[34mINFO[0m	info
	wantLog := `\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d\.\d{3}[+|-]\d{4}\s+.*INFO.*\s+info\s*$`

	l.Debug("debug")
	l.Info("info")

	b, err := os.ReadFile(f.Name())
	if err != nil {
		t.Fatal(err)
	}
	ok, err := regexp.MatchString(wantLog, string(b))
	if err != nil {
		t.Errorf("UnmarshalConfig(): MatchString failed %v", err)
	}
	if !ok {
		t.Errorf("UnmarshalConfig() sample log output: %q, want %q",
			string(b), wantLog)
	}

	// t.Error("intentional")
}


pkg/log/log.go
// Copyright (c) 2023 Remo Ronca 106963724+sobchak-security@users.noreply.github.com
// MIT License

package log

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

// configuration keys for enhanced encoders
const (
	ConfigKeyTimeEncoderInt64Seconds = "int64seconds"
	ConfigKeyTimeEncoderShort        = "short"
)

var (
	// Levels provide a convenient way to list all supported log level strings.
	Levels = []string{
		// omitting DPanic and Panic
		zapcore.DebugLevel.String(),
		zapcore.InfoLevel.String(),
		zapcore.WarnLevel.String(),
		zapcore.ErrorLevel.String(),
		zapcore.FatalLevel.String(),
	}
)

// ConfigWrapper is a simple unmarshaling wrapper of zap's Config structure. Let
// cfg be a variable of zap.Config, then, admittedly, the rather clumsy way of
// calling the unmarshaling function would be
// _ = (*ConfigWrapper)(&tt.cfg).Unmarshal(m)
// ... deal with it.
type ConfigWrapper zap.Config

// Unmarshal supports unmarshaling a zap logger configuration from a map. This
// way, in a pre-proccessing step, different configuration file formats can be
// parsed. Also, enhancements, like new encoders, are processed. Lastly,
// environment variables are resolved, including an attempt, at ensuring the
// variable HOSTNAME (not available on every platform), is made.
func (cfg *ConfigWrapper) Unmarshal(m map[string]interface{}) error {
	b, err := json.Marshal(m)
	if err != nil {
		return fmt.Errorf("UnmarshalConfig(): marshaling to JSON failed - %w", err)
	}
	if _, ok := os.LookupEnv("HOSTNAME"); !ok {
		var hostname string
		if hostname, err = os.Hostname(); err != nil || len(hostname) <= 0 {
			hostname, _ = os.LookupEnv("HOST")
		}
		if len(hostname) > 0 {
			os.Setenv("HOSTNAME", hostname)
		}
	}
	b = []byte(os.ExpandEnv(string(b)))
	if err := json.Unmarshal(b, cfg); err != nil {
		return fmt.Errorf("UnmarshalConfig(): unmarshaling (full) from JSON failed - %w", err)
	}

	var enhancements struct {
		EncoderConfig struct {
			TimeEncoder string
		}
	}
	if err := json.Unmarshal(b, &enhancements); err != nil {
		return fmt.Errorf("UnmarshalConfig(): unmarshaling (enhancements) from JSON failed - %w", err)
	}
	switch enhancements.EncoderConfig.TimeEncoder {
	case ConfigKeyTimeEncoderShort:
		cfg.EncoderConfig.EncodeTime = EpochShortTimeEncoder
	case ConfigKeyTimeEncoderInt64Seconds:
		cfg.EncoderConfig.EncodeTime = EpochInt64SecondsEncoder
	}
	return nil
}

// EpochInt64SecondsEncoder serializes a time.Time to a int64 number of seconds
// since the Unix epoch.
func EpochInt64SecondsEncoder(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
	enc.AppendInt64(t.Unix())
}

// EpochShortTimeEncoder serializes a time.Time to short time string.
func EpochShortTimeEncoder(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
	enc.AppendString(t.Format("06-01-02 15:04:05"))
}

// DevConfig returns a logger and atomic log level aimed at development
// environments with a highly opinionated configuration.
// NOTE this factory function panics if an error occurs.
func DevConfig() zap.Config {
	cfg := zap.NewDevelopmentConfig()
	cfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
	cfg.EncoderConfig.EncodeLevel = zapcore.CapitalLevelEncoder
	return cfg
}

// StdConfig returns a logger and atomic log level aimed at development
// environments with a highly opinionated configuration.
// NOTE this factory function panics if an error occurs.
func StdConfig() zap.Config {
	cfg := zap.NewProductionConfig()
	cfg.Encoding = "console"
	cfg.DisableCaller = true
	cfg.DisableStacktrace = true
	cfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
	cfg.EncoderConfig.EncodeLevel = zapcore.CapitalColorLevelEncoder
	cfg.EncoderConfig.EncodeDuration = zapcore.StringDurationEncoder
	return cfg
}
